compiler/src/iree/compiler/Codegen/Common/ConvertBf16ArithToF32.cpp

```diff
@@ -0,0 +1,312 @@
+    this->addArgumentMaterialization(convertRankedFloat);
+    this->addSourceMaterialization(convertRankedFloat);
+    this->addTargetMaterialization(convertRankedFloat);
+  }
+};
+
+// Tries to completely convert a generic Operation.
+// This will process attributes, result types, and nested regions.
+struct GenericTypeConversionPattern : public ConversionPattern {
+  GenericTypeConversionPattern(MLIRContext *context,
+                               TypeConverter &typeConverter)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), 0, context) {}
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    // Convert attributes only if this is a constant-like op.
+    // This is because some ops use typed attributes for structural information
+    // - like linalg ops using i64 for dimension indices - and if we converted
+    // them all the ops would become invalid. This may still be too broad,
+    // though, if some constant ops include attributes with both the type we
+    // want to convert and structural information in the same type.
+    llvm::SmallVector<NamedAttribute> newAttrs;
+    if (op->hasTrait<OpTrait::ConstantLike>() ||
+        isa<IREE::Util::GlobalOpInterface>(op)) {
+      for (auto attr : op->getAttrs()) {
+        auto newAttr = convertAttribute(op->getLoc(), attr.getValue(),
+                                        *getTypeConverter());
+        newAttrs.push_back(NamedAttribute(attr.getName(), newAttr));
+      }
+    } else {
+      newAttrs.append(op->getAttrs().begin(), op->getAttrs().end());
+    }
+
+    llvm::SmallVector<Type, 4> newResults;
+    (void)getTypeConverter()->convertTypes(op->getResultTypes(), newResults);
+
+    OperationState state(op->getLoc(), op->getName().getStringRef(), operands,
+                         newResults, newAttrs, op->getSuccessors());
+
+    for (Region &r : op->getRegions()) {
+      Region *newRegion = state.addRegion();
+      rewriter.inlineRegionBefore(r, *newRegion, newRegion->begin());
+      TypeConverter::SignatureConversion result(newRegion->getNumArguments());
+      (void)getTypeConverter()->convertSignatureArgs(
+          newRegion->getArgumentTypes(), result);
+      rewriter.applySignatureConversion(newRegion, result);
+    }
+
+    Operation *newOp = rewriter.create(state);
+    rewriter.replaceOp(op, newOp->getResults());
+    return success();
+  }
+};
+
+template <typename OpTy, typename TypeTy,
+          typename OperandToResultWidthLegalityRelation>
+struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, typename OpTy::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto resultType =
+        this->getTypeConverter()->convertType(op.getResult().getType());
+    auto operandType =
```

Comment:
If I am not mistaken this is just the type of the operand obtained from the `adaptor`.

Reply:
Done.